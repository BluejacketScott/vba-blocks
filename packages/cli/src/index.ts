export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
