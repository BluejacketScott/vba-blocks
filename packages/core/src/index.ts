export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest
} from './manifest';
export { Config } from './config';
