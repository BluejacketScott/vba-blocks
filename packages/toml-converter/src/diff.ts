import { diff as deepDiff, Index, Path } from 'deep-diff';
import traverse from 'traverse';
import clone from 'clone';

/*
  deep-diff has an admittedly simple approach to handling array changes,
  this adds a system for better determining array moves
  
  1. Walk lhs object and identify arrays
  2. Find removed, added, and moved values
  3. Store change information and apply move/remove operations
  4. Diff modified object to identify remaining changes

  Adds a new kind of change: 'R' (for reorder)

  ```js
  {
    kind: 'R',
    path: ['path', 'to', 'array'],
    indices: [2,1,0]
  }
  ```

  Above = Reorder from [0,1,2] to [2,1,0]
*/

export interface Change {
  kind: 'N' | 'D' | 'E' | 'A' | 'R';
  path?: Path;
  lhs?: any;
  rhs?: any;
  index?: number;
  indices?: number[];
  item?: Change;
}

// There are probably heuristics to identify both moves in the array and object changes,
// but for simplicity an id function is provided to identify values in the array
export type getObjectId = (value: any, path: Path, found_in: any[]) => any;

export default function diff(
  lhs: any,
  rhs: any,
  options: { getId?: getObjectId } = {}
): Change[] {
  const array_changes = diffArray(lhs, rhs, options);
  lhs = applyArrayChanges(lhs, array_changes);

  const deep_changes = deepDiff(lhs, rhs) || [];
  return [...array_changes, ...deep_changes];
}

/*
  [1,2,3,4,5] to [6,4,3,2,6,7]

  1. Remove - [2,3,4] (at index)
  2. Reorder - [4,3,2] (by index)
  3. Add - [4,3,2,6,7] (at index)
*/
function diffArray(
  lhs: any,
  rhs: any,
  options: { getId?: getObjectId }
): Change[] {
  const { getId = (value: any) => value } = options;

  const paths: Path[] = traverse(lhs).paths();
  let changes: Change[] = [];

  for (const path of paths) {
    const previous = get(lhs, path);
    const next = get(rhs, path);

    if (!next || !Array.isArray(previous)) continue;

    const previous_ids = previous.map((value: any) =>
      getId(value, path, previous)
    );
    const next_ids = next.map((value: any) => getId(value, path, next));

    const remove = indexedDifference(previous_ids, next_ids);
    const add = indexedDifference(next_ids, previous_ids);
    const reorder = reordered(
      removeByIndex(previous_ids, remove),
      removeByIndex(next_ids, add)
    );

    remove.forEach(index => changes.push({ kind: 'D', path, index }));
    if (reorder.length) changes.push({ kind: 'R', path, indices: reorder });
    add.forEach(index =>
      changes.push({
        kind: 'A',
        path,
        index,
        item: { kind: 'N', rhs: next[index] }
      })
    );
  }

  return changes;
}

function applyArrayChanges(value: any, changes: Change[]): any {
  value = clone(value);

  for (const change of changes) {
    const { kind, path, index, indices, item } = change;
    const array = get(value, change.path!);
    if (kind === 'D') {
      array.splice(index, 1);
    } else if (kind === 'A') {
      array.splice(index, 0, item!.rhs);
    } else if (kind === 'R') {
      const original = array.slice();
      indices!.forEach((to, from) => (array[to] = original[from]));
    }
  }

  return value;
}

function difference(values: any[], reference: any[]): any[] {
  return values.filter((value: any) => !reference.includes(value));
}

function indexedDifference(values: any[], reference: any[]): number[] {
  return values.reduce((memo: number[], value: any, index: number) => {
    if (!reference.includes(value)) memo.push(index);
    return memo;
  }, []);
}

function removeByIndex(values: any[], to_remove: number[]): any[] {
  return values.filter(
    (value: any, index: number) => !to_remove.includes(index)
  );
}

function reordered(from: any[], to: any[]): number[] {
  const found: boolean[] = [];
  const reorder = from.map((value: any) => {
    let index = to.indexOf(value);
    while (index >= 0 && found[index]) {
      index = to.indexOf(value, index + 1);
    }

    if (index < 0)
      throw new Error(`Could not find moved value "${value}" in destination`);
    found[index] = true;

    return index;
  });

  const noop = reorder.every((to, from) => to === from);

  return noop ? [] : reorder;
}

function get(object: any, path: Path): any {
  for (const part of path) {
    if (!object) return undefined;
    object = object[part];
  }
  return object;
}
