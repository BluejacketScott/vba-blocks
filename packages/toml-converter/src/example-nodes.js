// TEMP

[
  {
    type: 'ObjectPath',
    value: ['package'],
    line: 1,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: 'package-name',
      line: 2,
      column: 8
    },
    line: 2,
    column: 1,
    key: 'name'
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '0.0.0',
      line: 3,
      column: 11
    },
    line: 3,
    column: 1,
    key: 'version'
  },
  {
    type: 'Assign',
    value: {
      type: 'Array',
      value: [
        {
          type: 'String',
          value: 'Author 1',
          line: 4,
          column: 12
        }
      ],
      line: 4,
      column: 11
    },
    line: 4,
    column: 1,
    key: 'authors'
  },
  {
    type: 'ObjectPath',
    value: ['dependencies'],
    line: 12,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '1.0.0',
      line: 13,
      column: 5
    },
    line: 13,
    column: 1,
    key: 'a'
  },
  {
    type: 'Assign',
    value: {
      type: 'InlineTable',
      value: [
        {
          type: 'InlineTableValue',
          value: {
            type: 'String',
            value: '...git',
            line: 14,
            column: 13
          },
          line: 14,
          column: 7,
          key: 'git'
        },
        {
          type: 'InlineTableValue',
          value: {
            type: 'String',
            value: 'weird',
            line: 14,
            column: 38
          },
          line: 14,
          column: 25,
          key: 'spacing'
        }
      ],
      line: 14,
      column: 5
    },
    line: 14,
    column: 1,
    key: 'b'
  },
  {
    type: 'ObjectPath',
    value: ['dependencies', 'c'],
    line: 16,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '...path',
      line: 17,
      column: 8
    },
    line: 17,
    column: 1,
    key: 'path'
  },
  {
    type: 'ArrayPath',
    value: ['array'],
    line: 19,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'Integer',
      value: 0,
      line: 20,
      column: 9
    },
    line: 20,
    column: 1,
    key: 'index'
  },
  {
    type: 'ArrayPath',
    value: ['array'],
    line: 22,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'Integer',
      value: 1,
      line: 23,
      column: 9
    },
    line: 23,
    column: 1,
    key: 'index'
  }
];
