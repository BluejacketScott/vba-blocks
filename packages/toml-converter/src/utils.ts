export function toLines(value: string): string[] {
  return value.replace(/\r/g, '').split('\n');
}
