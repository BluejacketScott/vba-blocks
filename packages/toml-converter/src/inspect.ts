import { Path } from 'deep-diff';
import { graceful as detectNewline } from 'detect-newline';
import { Change } from './diff';

const { parse: inspectToml } = require('toml/lib/parser');
const { compile: compileToml } = require('toml/lib/compiler');

export interface Node {
  type:
    | 'ObjectPath'
    | 'ArrayPath'
    | 'Assign'
    | 'String'
    | 'Float'
    | 'Integer'
    | 'Boolean'
    | 'Array'
    | 'InlineTable'
    | 'InlineTableValue'
    | 'Date';
  line: number;
  column: number;
  value: any;
  key?: string;
}

export default function inspect(toml: string): AST {
  const nodes: Node[] = inspectToml(toml);
  const landmarks = toLandmarks(nodes);

  // TODO
  return {};
}

// TODO Unify landmarks and AST

export interface AST {}

export type Landmarks = { [key: string]: Landmark } | Landmark[];
export interface Landmark {
  node: Node;
  inline?: Node;
  children?: Landmarks;
}

export function toLandmarks(nodes: Node[]): Landmarks | Landmark {
  const landmarks = {};
  let active = landmarks;

  for (const node of nodes) {
    // For inline tables and arrays, need to store some additional information
    // store in `inline` value
    const inline =
      node.value.type === 'InlineTable' || node.value.type === 'Array'
        ? node.value
        : null;

    switch (node.type) {
      case 'ObjectPath':
        // { "value": ["package"] }
        active = {};
        set(landmarks, node.value, { node, children: active });
        break;
      case 'ArrayPath':
        // { "value": ["array"] }
        active = {};
        push(landmarks, node.value, { node, children: active });
        break;
      case 'Assign':
      case 'InlineTableValue':
        // { "key": "name", "value": { "type": "String", "value": "package-name" }
        const children = toLandmarks([node.value]);
        set(
          active,
          [node.key!],
          inline ? { node, inline, children } : { node, children }
        );
        break;

      case 'InlineTable':
        return toLandmarks(node.value);
      case 'Array':
        return node.value.map((value: any) => toLandmarks([value]));

      case 'String':
      case 'Float':
      case 'Integer':
      case 'Boolean':
      case 'Date':
        return { node };

      default:
        throw new Error(`Unrecognized node landmark type "${node.type}"`);
    }
  }

  return landmarks;
}

function get(object: any, path: Path): any {
  if (!object) return undefined;
  object = object[path[0]];

  for (const part of path.slice(1)) {
    if (!object || !object.children) return undefined;
    object = object.children[part];
  }
  return object;
}

export function set(object: any, path: Path, value: any) {
  if (path.length === 1) {
    object[path[0]] = value;
  } else {
    const parent = get(object, path.slice(0, -1));
    if (!parent || !parent.children)
      throw new Error(`Could not find path ${path} in object`);

    const key = path[path.length - 1];
    parent.children[key] = value;
  }

  return value;
}

export function push(object: any, path: Path, value: any) {
  const array = get(object, path) || set(object, path, []);
  array.push(value);
}
