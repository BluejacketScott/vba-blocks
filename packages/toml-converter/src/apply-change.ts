import { AST } from './inspect';
import { Change } from './diff';

export default function applyChange(
  toml: string[],
  ast: AST,
  change: Change
): string[] {
  // TODO
  return toml;
}
