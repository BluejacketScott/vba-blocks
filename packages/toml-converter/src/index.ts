import { parse as parseToml } from 'toml';
import { graceful as detectNewline } from 'detect-newline';
import inspect, { Node, AST } from './inspect';
import diff, { Change, getObjectId } from './diff';
import applyChange from './apply-change';
import { toLines } from './utils';

export { Node, AST, Change, getObjectId };

export interface Options {
  getId?: getObjectId;
  trailingComma?: boolean;
}

export function parse(toml: string): any {
  return parseToml(toml);
}

export function convert(value: any, options: Options = {}): string {
  return patch('', value, options);
}

export function patch(toml: string, value: any, options: Options = {}): string {
  const lines = toLines(toml);
  const newline = detectNewline(toml);

  const existing = parse(toml);
  const changes = diff(existing, value, options);

  // applyChange works directly on lines and does not update AST as-needed
  // meaning the AST has to be re-calculated for each change
  //
  // This is not ideal, but the assumption is that patches will be small and infrequent
  const result = changes.reduce((lines, change) => {
    const ast = inspect(lines.join(newline));
    return applyChange(lines, ast, change);
  }, lines);

  return result.join(newline);
}
