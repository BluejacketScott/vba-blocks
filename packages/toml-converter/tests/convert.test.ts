import { parse, convert } from '../src';
import { fixture } from './patch.test';

test('convert', () => {
  expect(convert(parse(fixture))).toMatchSnapshot();
});
