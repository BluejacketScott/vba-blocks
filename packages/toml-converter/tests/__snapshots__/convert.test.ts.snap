// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert 1`] = `""`;
